src/message_type.rs

```diff
@@ -17,7 +17,152 @@
+        match &self.message {
+            PbftMessageWrapper::PbftMessage(m) => &m.get_info(),
+            PbftMessageWrapper::PbftViewChange(m) => &m.get_info(),
+        }
+    }
+
+    pub fn set_info(&mut self, info: PbftMessageInfo) {
+        match self.message {
+            PbftMessageWrapper::PbftMessage(ref mut m) => m.set_info(info),
+            PbftMessageWrapper::PbftViewChange(ref mut m) => m.set_info(info),
+        }
+    }
+
+    pub fn get_block(&self) -> &PbftBlock {
+        match &self.message {
+            PbftMessageWrapper::PbftMessage(m) => m.get_block(),
+            PbftMessageWrapper::PbftViewChange(_) => {
+                panic!("ParsedPeerMessage.get_block found a view change message!")
+            }
+        }
+    }
+
+    pub fn get_pbft_message(&self) -> &PbftMessage {
+        match &self.message {
+            PbftMessageWrapper::PbftMessage(m) => m,
+            PbftMessageWrapper::PbftViewChange(_) => {
+                panic!("ParsedPeerMessage.get_pbft_message found a view change message!")
+            }
+        }
+    }
+
+    pub fn get_view_change_message(&self) -> &PbftViewChange {
+        match &self.message {
+            PbftMessageWrapper::PbftMessage(_) => {
+                panic!("ParsedPeerMessage.get_view_change_message found a pbft message!")
+            }
+            PbftMessageWrapper::PbftViewChange(m) => m,
+        }
+    }
+
+    pub fn from_peer_message(message: PeerMessage) -> Result<Self, PbftError> {
+        verify_peer_message_hash(&message.content, &message.header.content_sha512)?;
+
+        let parsed_message = match protobuf::parse_from_bytes::<PbftMessage>(&message.content) {
+            Ok(m) => PbftMessageWrapper::PbftMessage(m),
+            Err(_) => match protobuf::parse_from_bytes::<PbftViewChange>(&message.content) {
+                Ok(m) => PbftMessageWrapper::PbftViewChange(m),
+                Err(_) => return Err(PbftError::InternalError("Couldn't parse message!".into())),
+            },
+        };
+
+        Ok(Self {
+            header_bytes: message.header_bytes,
+            header_signature: message.header_signature,
+            message: parsed_message,
+            message_bytes: message.content.clone(),
+        })
+    }
+
+    pub fn from_bytes(message: Vec<u8>) -> Result<Self, PbftError> {
+        let peer_message = PeerMessage {
+            content: message,
+            header_bytes: b"myself".to_vec(),
+            ..Default::default()
+        };
+
+        Self::from_peer_message(peer_message)
+    }
+}
+
+/// Verifies that the SHA-512 hash of the given content matches the given hash
+pub fn verify_peer_message_hash(content: &[u8], content_hash: &[u8]) -> Result<(), PbftError> {
+    let mut sha = Sha512::new();
+    sha.input(&content);
+    let computed_sha512 = hex::decode(sha.result_str()).unwrap();
+
+    if content_hash.is_empty() {
+        warn!("Skipping hash verification due to empty content hash!");
```

Comment:
This is happening due to us constructing messages that the nodes send to themselves, and we shouldn't be constructing `Consensus*` protobuf messages from within PBFT, so e.g. the header_signature is just an empty vec of bytes. This will hopefully go away, depending on the results of that discussion.